src/hal/src/lib.rs

```diff
@@ -467,7 +486,64 @@ pub struct Limits {
 
     /// The alignment of the vertex buffer stride.
     pub min_vertex_input_binding_stride_alignment: buffer::Offset,
+}
 
+/// Resource limits related to Descriptor Indexing.
+#[derive(Clone, Copy, Debug, Default, PartialEq)]
+#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
+pub struct DescriptorIndexingCapabilities {
+    ///
+    pub max_update_after_bind_descriptors_in_all_pools: u32,
+    ///
+    pub shader_uniform_buffer_array_non_uniform_indexing_native: bool,
+    ///
+    pub shader_sampled_image_array_non_uniform_indexing_native: bool,
+    ///
+    pub shader_storage_buffer_array_non_uniform_indexing_native: bool,
+    ///
+    pub shader_storage_image_array_non_uniform_indexing_native: bool,
+    ///
+    pub shader_input_attachment_array_non_uniform_indexing_native: bool,
+    ///
+    pub robust_buffer_access_update_after_bind: bool,
+    ///
+    pub quad_divergent_implicit_lod: bool,
+    ///
+    pub max_per_stage_descriptor_update_after_bind_samplers: u32,
+    ///
+    pub max_per_stage_descriptor_update_after_bind_uniform_buffers: u32,
+    ///
+    pub max_per_stage_descriptor_update_after_bind_storage_buffers: u32,
+    ///
+    pub max_per_stage_descriptor_update_after_bind_sampled_images: u32,
+    ///
+    pub max_per_stage_descriptor_update_after_bind_storage_images: u32,
+    ///
+    pub max_per_stage_descriptor_update_after_bind_input_attachments: u32,
+    ///
+    pub max_per_stage_update_after_bind_resources: u32,
+    ///
+    pub max_descriptor_set_update_after_bind_samplers: u32,
+    ///
+    pub max_descriptor_set_update_after_bind_uniform_buffers: u32,
+    ///
+    pub max_descriptor_set_update_after_bind_uniform_buffers_dynamic: u32,
```

Comment:
we have quite a few places for `max_XXX_descriptor_XXX_some_stage`. I wonder if we can make this nicer? I.e.
```rust
struct BindLimits {
  max_uniform_buffers: u32,
  max_samplers: u32,
  ...
}
pub per_stage_descriptor_update_limits: BindLimits,
pub descriptor set_update_limits: BindLimits,
```
We'd probably use it in the `Limits` too!

Reply:
I think we'll need to get rid of the `*_update_after_bind_*` entries here, since `gfx` doesn't (yet?) support that.

I also noticed most of the core `Limits` are `usize` when the actual Vulkan value is `u32`, is this intentional? And should I follow that pattern for mesh shader and descriptor indexing? The best reason I can come up with is that these values will likely be used in calculations to size rust vecs or arrays, which would operate on `usize`.

Reply:
Not intentional, we should probably change them to `u32`.

Reply:
So are we removing the `_update_after_bind_` for now?

Reply:
Admittedly I'm not sure what we'd need for this feature. I've removed the update-after-bind stuff because gfx itself doesn't support that feature. I believe update-after-bind is pretty important for descriptor indexing--although, again, I could be wrong here.

@cwfitzgerald I recall we needed descriptor indexing to handle a bug in Intel drivers for WebGPU... can I defer to you for what is important to expose as a part of `hal` for descriptor indexing? Were you imagining this being a mostly "private capabilities"-type feature or something publicly exposed?

Reply:
this is the stuff from here:
> VK_DESCRIPTOR_BINDING_UPDATE_AFTER_BIND_BIT indicates that if descriptors in this binding are updated between when the descriptor set is bound in a command buffer and when that command buffer is submitted to a queue, then the submission will use the most recently set descriptors for this binding and the updates do not invalidate the command buffer. Descriptor bindings created with this flag are also partially exempt from the external synchronization requirement in vkUpdateDescriptorSetWithTemplateKHR and vkUpdateDescriptorSets. Multiple descriptors with this flag set can be updated concurrently in different threads, though the same descriptor must not be updated concurrently by two threads. Descriptors with this flag set can be updated concurrently with the set being bound to a command buffer in another thread, but not concurrently with the set being reset or freed.

We don't have that supported yet, so we can leave this off PR.
However, it would be nice to at least group those descriptor limits together in a struct, so that if/when we need to add another set of limits for the "update_after_bind", we can just re-use the same struct type.

Reply:
I'll do another pass to see what I can group together... I remember the last time I looked at it (briefly!) every group of limits were just slightly different from another and didn't meaningfully have an abstraction to pull out.

I'd also like to wait on @cwfitzgerald just to make sure this is sufficient for the Intel/WebGPU work?

Reply:
Also removed the _update_after_bind_ fields

Reply:
So for update after bind, the idea is that it actually introduces no more requirements on the API to just use update after bind descriptors like normal descriptors. i.e. update after bind descriptors are a superset of non-update after bind descriptors, so we should just implicitly use them to raise the limits on intel.